barretenberg/cpp/src/barretenberg/stdlib/primitives/group/cycle_group.cpp

```diff
@@ -340,74 +337,67 @@ template <typename Builder> void cycle_group<Builder>::standardize()
 }
 
 /**
- * @brief Evaluates a doubling. Uses Ultra double gate
+ * @brief Evaluates a point doubling using Ultra ECC double gate (if non-constant)
  *
  * @tparam Builder
- * @param hint : value of output point witness, if known ahead of time (used to avoid modular inversions during witgen)
- * @return cycle_group<Builder>
+ * @param hint native result of the doubling (optional; used to avoid modular inversions during witgen)
+ * @return cycle_group<Builder> The doubled point
  */
 template <typename Builder>
 cycle_group<Builder> cycle_group<Builder>::dbl(const std::optional<AffineElement> hint) const
-    requires IsUltraArithmetic<Builder>
 {
-    // ensure we use a value of y that is not zero. (only happens if point at infinity)
-    // this costs 0 gates if `is_infinity` is a circuit constant
-    auto modified_y = field_t::conditional_assign(is_point_at_infinity(), 1, y).normalize();
-
-    // We have to return the point at infinity immediately
-    // Cause in that very case the `modified_y` is a constant value, with witness_index = -1
-    // Hence the following `create_ecc_dbl_gate` will throw an ASSERTION error
-    if (this->is_point_at_infinity().is_constant() && this->is_point_at_infinity().get_value()) {
+    // If this is a constant point at infinity, return early
+    if (this->is_constant_point_at_infinity()) {
         return *this;
     }
 
-    cycle_group result;
-    if (hint.has_value()) {
-        const bb::fr x3 = hint.value().x;
-        const bb::fr y3 = hint.value().y;
-        if (is_constant()) {
-            result = cycle_group(x3, y3, is_point_at_infinity());
-            // We need to manually propagate the origin tag
-            result.set_origin_tag(get_origin_tag());
-
-            return result;
-        }
+    // To support the point at infinity, we conditionally modify y to be 1 to avoid division by zero in the
+    // doubling formula
+    auto modified_y = field_t::conditional_assign(is_point_at_infinity(), 1, y).normalize();
 
-        result = cycle_group(witness_t(context, x3), witness_t(context, y3), is_point_at_infinity());
+    // Compute the doubled point coordinates (either from hint or by native calculation)
+    bb::fr x3;
+    bb::fr y3;
+    if (hint.has_value()) {
```

Comment:
The remaining changes here are simply to make the logic easier to follow. Rather than having nested conditionals and early returns we just handle two distinct phases separately: (1) get the native result of the doubling (possibly from hint), (2) construct the witnesses + constraint (unless constant).